miqa/core/tests/test_rest.py

```diff
@@ -0,0 +1,178 @@
+from uuid import UUID
+
+import pytest
+
+from .fuzzy import PATH_RE, TIMESTAMP_RE
+
+
+def experiment_to_dict(session):
+    def a(experiment):
+        return {
+            'id': experiment.id,
+            'name': experiment.name,
+            'note': experiment.note,
+            'session': {'id': session.id, 'name': session.name},
+        }
+
+    return a
+
+
+def scan_to_dict(experiment, site):
+    def a(scan):
+        return {
+            'id': scan.id,
+            'scan_id': scan.scan_id,
+            'scan_type': scan.scan_type,
+            'decision': scan.decision,
+            'notes': [],
+            'site': UUID(site.id),
+            'experiment': UUID(experiment.id),
+        }
+
+    return a
+
+
+def image_to_dict(scan):
+    def a(image):
+        return {'id': image.id, 'name': image.name, 'scan': UUID(scan.id)}
+
+    return a
+
+
+def note_to_dict(user):
+    def a(note):
+        return {
+            'id': note.id,
+            'creator': {'first_name': user.first_name, 'last_name': user.last_name},
+            'note': note.note,
+            'created': TIMESTAMP_RE,
+            'modified': TIMESTAMP_RE,
+        }
+
+    return a
+
+
+def compare_lst(lst1, lst2, to_dict):
+    lst1 = list(map(lambda x: to_dict(x), lst1))
+
+    diff = [i for i in lst1 + lst2 if i not in lst1 or i not in lst2]
```

Comment:
This seems like it would be more readable as set-based operations.

```python
from typing import Callable, List

def compare_list(models: List[Model], dicts: List[dict], to_dict: Callable[[Model], dict]) -> bool:
    set1 = set(map(to_dict, lst1))
    return set1 == set(lst2)
```

Reply:
This would be nice, but in python you can't create a set of dictionaries:

```
>>> lst = [{"test": "test"}, {"test": "test"}]
>>> set(lst)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'dict'
```

Reply:
Good point, the implementation details of `set` are leaking. We can get a hashable form of the dictionary via `tuple(the_dict.items())`, but maybe it's not worth it. We'll punt on this now and can always revisit later.

Reply:
I even tried doing something like this:
```
set1 = set(tuple(sorted(to_dict(d).items())) for d in models)
set2 = set(tuple(sorted(d.items())) for d in dicts)
return set1 == set2
```
But in some of the comparisons, I am using regular expressions to match and those are also not hashable

Reply:
Was there a reason the built-in pytest dictionary diffing (just using `==`) was insufficient?

Reply:
Yes, I was trying to compare two lists of dictionaries (comparing database models to the response from the rest API). However, because I had lists, using `==` would take order of items into consideration which is not desired. As mentioned above, I was not able to use sets because dictionaries are not hash-able in python.

I guess using `==` would work if we sorted both lists beforehand